File: docs/vsto/codesnippet/VisualBasic/my excel chart/ThisWorkbook.vb

Public Class ThisWorkbook

    '<Snippet1>
    ' Declare the menu variable at the class level.
    Private WithEvents menuCommand As Office.CommandBarButton
    Private menuTag As String = "A unique tag"


    ' Call AddMenu from the Startup event of ThisWorkbook.
    Private Sub ThisWorkbook_Startup(ByVal sender As Object, ByVal e As System.EventArgs) _
        Handles Me.Startup

        CheckIfMenuBarExists()
        AddMenuBar()
    End Sub


    ' If the menu already exists, remove it.
    Private Sub CheckIfMenuBarExists()
        Try
            Dim foundMenu As Office.CommandBarPopup = _
                Me.Application.CommandBars.ActiveMenuBar.FindControl( _
                    Office.MsoControlType.msoControlPopup, System.Type.Missing, menuTag, True, True)

            If foundMenu IsNot Nothing Then
                foundMenu.Delete(True)
            End If

        Catch ex As Exception
            MessageBox.Show(ex.Message)
        End Try
    End Sub


    ' Create the menu, if it does not exist.
    Private Sub AddMenuBar()

        Try
            Dim menuBar As Office.CommandBar = Application.comm.CommandBars.ActiveMenuBar
            Dim menuCaption As String = "Ne&w Menu"

            If menuBar IsNot Nothing Then
                Dim cmdBarControl As Office.CommandBarPopup = Nothing
                Dim controlCount As Integer = menuBar.Controls.Count

                ' Add the new menu.
                cmdBarControl = CType(menuBar.Controls.Add( _
                    Type:=Office.MsoControlType.msoControlPopup, Before:=controlCount, Temporary:=True),  _
                    Office.CommandBarPopup)

                cmdBarControl.Caption = menuCaption
                cmdBarControl.Tag = menuTag

                ' Add the menu command.
                menuCommand = CType(cmdBarControl.Controls.Add( _
                    Type:=Office.MsoControlType.msoControlButton, Temporary:=True),  _
                    Office.CommandBarButton)

                With menuCommand
                    .Caption = "&New Menu Command"
                    .Tag = "NewMenuCommand"
                    .FaceId = 65
                End With
            End If

        Catch ex As Exception
            MessageBox.Show(ex.Message)
        End Try
    End Sub


    ' Add text to cell A1 when the menu is clicked.
    Private Sub menuCommand_Click(ByVal Ctrl As Microsoft.Office.Core.CommandBarButton, _
        ByRef CancelDefault As Boolean) Handles menuCommand.Click

        Globals.Sheet1.Range("A1").Value2 = "The menu command was clicked."
    End Sub
    '</Snippet1>

End Class

File: docs/vsto/codesnippet/CSharp/Trin_VstcoreProgrammingControlsExcelCS/ThisWorkbook.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Xml.Linq;
using Microsoft.Office.Tools.Excel;
using Microsoft.VisualStudio.Tools.Applications.Runtime;
using Excel = Microsoft.Office.Interop.Excel;
using Office = Microsoft.Office.Core;

namespace Trin_VstcoreProgrammingControlsExcelCS
{
    public partial class ThisWorkbook
    {
        //<Snippet1>
        // Declare the menu variable at the class level.
        private Office.CommandBarButton menuCommand;
        private string menuTag = "A unique tag";

        private void ThisWorkbook_Startup(object sender, System.EventArgs e)
        {
            CheckIfMenuBarExists();
            AddMenuBar();
        }

        // If the menu already exists, remove it.
        private void CheckIfMenuBarExists()
        {
            try
            {
                Office.CommandBarPopup foundMenu = (Office.CommandBarPopup)
                    this.Application.CommandBars.ActiveMenuBar.FindControl(
                    Office.MsoControlType.msoControlPopup, menuTag, true, true);

                if (foundMenu != null)
                {
                    foundMenu.Delete(true);
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }
        }

        // Create the menu, if it does not exist.
        private void AddMenuBar()
        {
            try
            {
                Office.CommandBarPopup cmdBarControl = null;
                Office.CommandBar menubar = (Office.CommandBar)Application.CommandBars.ActiveMenuBar;
                int controlCount = menubar.Controls.Count;
                string menuCaption = "&New Menu";

                // Add the menu.
                cmdBarControl = (Office.CommandBarPopup)menubar.Controls.Add(
                    Office.MsoControlType.msoControlPopup, controlCount, true);

                if (cmdBarControl != null)
                {
                    cmdBarControl.Caption = menuCaption;
                    cmdBarControl.Tag = menuTag;

                    // Add the menu command.
                    menuCommand = (Office.CommandBarButton)cmdBarControl.Controls.Add(
                        Office.MsoControlType.msoControlButton, true);

                    menuCommand.Caption = "&New Menu Command";
                    menuCommand.Tag = "NewMenuCommand";
                    menuCommand.FaceId = 65;

                    menuCommand.Click += new Microsoft.Office.Core._CommandBarButtonEvents_ClickEventHandler(
                        menuCommand_Click);
                }
            }
            catch (Exception e)
            {
                MessageBox.Show(e.Message);
            }
        }

        // Add text to cell A1 when the menu is clicked.
        private void menuCommand_Click(Microsoft.Office.Core.CommandBarButton Ctrl, ref bool CancelDefault)
        {
            Globals.Sheet1.Range["A1"].Value2 = "The menu command was clicked.";
        }
        //</Snippet1>

        private void ThisWorkbook_Shutdown(object sender, System.EventArgs e)
        {
        }

        #region VSTO Designer generated code

        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InternalStartup()
        {
            this.Startup += new System.EventHandler(ThisWorkbook_Startup);
            this.Shutdown += new System.EventHandler(ThisWorkbook_Shutdown);
        }

        #endregion

    }
}
